icechunk-python/tests/test_stateful_parallel_sessions.py

```diff
@@ -0,0 +1,232 @@
+            arr = zarr.open_array(path=array, store=store)
+            arr[slicers] = new_data
+
+        self.execute_on_parallel(data=data, func=write)
+        self.has_changes = True
+
+    @precondition(lambda self: bool(self.all_arrays))
+    @rule(data=st.data())
+    def delete_chunk(self, data: st.DataObject) -> None:
+        array = data.draw(st.sampled_from(sorted(self.all_arrays)))
+        arr = zarr.open_array(path=array, store=self.serial.store)
+        chunk_path = data.draw(icst.chunk_paths(numblocks=arr.cdata_shape))
+        path = f"{array}/c/{chunk_path}"
+        note(f"deleting chunk {path=!r}")
+        SyncStoreWrapper(self.serial.store).delete(path)
+        self.execute_on_parallel(
+            data=data, func=lambda store: SyncStoreWrapper(store).delete(path)
+        )
+        self.has_changes = True
+
+    def execute_on_parallel(
+        self, *, data: st.DataObject, func: Callable[..., None]
+    ) -> None:
+        """
+        Chooses one of self.parallel, self.fork1, or self.fork2
+        as the session on which to make changes using `func`.
+        """
+        if self.has_forks():
+            # prioritize drawing a fork first
+            name, session = data.draw(
+                st.sampled_from(
+                    [
+                        ("fork1", self.fork1),
+                        ("parallel", self.parallel),
+                        ("fork2", self.fork2),
+                    ]
+                )
+            )
+        else:
+            name, session = "parallel", self.parallel
+        note(f"executing on {name}")
+        assert session is not None
+        func(session.store)
+
+    @precondition(lambda self: not self.has_forks())
+    @rule()
+    def fork_pickle(self) -> None:
+        note("forking with pickle")
+        with self.parallel.allow_pickling():
+            self.fork1 = pickle.loads(pickle.dumps(self.parallel))
+            self.fork2 = pickle.loads(pickle.dumps(self.parallel))
+
+    @precondition(lambda self: not self.has_forks())
+    @rule()
+    def fork_threads(self) -> None:
+        note("forking with reference (threads)")
+        self.fork1 = self.parallel
+        self.fork2 = self.parallel
+
+    @precondition(lambda self: self.has_forks())
+    @rule(two_to_one=st.booleans())
+    def merge(self, two_to_one: bool) -> None:
+        assert self.fork1 is not None
+        assert self.fork2 is not None
+        if two_to_one:
+            note("merging forks to base session, merging 2→1→parallel")
+            merge_sessions(self.fork1, self.fork2)
+            merge_sessions(self.parallel, self.fork1)
+        else:
+            note("merging forks to base session, merging 1→2→parallel")
+            merge_sessions(self.fork2, self.fork1)
+            merge_sessions(self.parallel, self.fork2)
+
+        self.fork1 = None
+        self.fork2 = None
+
+    @precondition(lambda self: not self.has_forks() and self.has_changes)
```

Comment:
this is basically saying we expect the user to fork, merge, and then commit the "base" `parallel` session